File: fact.ml
(*
    @@@ get_version: ocamlopt -version
    @@@ before: ocamlopt -o fact fact.ml
    @@@ instead: ./fact
    @@@ after: rm fact fact.cmi fact.cmx fact.o
    @@@ is_fast: 1
*)

open Printf;;

let times = 150000000
let fact_16 = 20922789888000

let rec fact n =
    match n with
    0 -> 1
    | _ -> n * fact(n - 1)
;;

let ok = ref true;;
for i = 1 to times do
    ok := !ok && fact(16) == fact_16
done;;

let str_result = if !ok then "ok" else "fail" in
Printf.printf "OCaml finish %d - %s\n" times str_result;;

File: fact.zig
// @@@ get_version: zig version
// @@@ before: zig build-exe fact.zig
// @@@ instead: ./fact 2>&1
// @@@ after: rm fact
// @@@ is_fast: 1

const std = @import("std");

const times: i64 = 100000000;
const fact_16: i64 = 20922789888000;

fn fact(n: i32) i64 {
    if (n == 1)
        return 1;
    return @as(i64, n) * fact(n - 1);
}

pub fn main() !void {
    const n: i32 = 16;
    var ok: bool = true;
    var i: i64 = 0;
    
    while (i < times) : (i += 1) {
        ok = ok and (fact(n) == fact_16);
    }

    std.debug.print("Zig finish {d} - {s}\n", .{ times, if (ok) "ok" else "fail" });
}
